File: annotation-service/utils/sts.js
/***
 * 
 * Copyright 2019-2021 VMware, Inc.
 * SPDX-License-Identifier: Apache-2.0
 * 
***/


const AWS = require('aws-sdk');
const config = require('../config/config');
const { AWSRESOURCE } = require('../config/constant')
const axios = require('axios')
const qs = require('qs')

async function prepareCredentials(resource, sessionDuration) {

    console.log(`[ STS ] Utils prepareCredentials`);
    if (config.ESP) {
        return prepareESPCredentails(resource, sessionDuration);
    } else {
        return prepareOSCredentails(resource, sessionDuration);
    }
}

async function prepareOSCredentails(resource, sessionDuration) {

    let roleArn = "";
    if (resource === AWSRESOURCE.S3) {
        roleArn = config.s3RoleArn;
    } else if (resource === AWSRESOURCE.SQS) {
        roleArn = config.sqsRoleArn;
    }

    let sts = new AWS.STS({
        endpoint: `https://sts.${config.region}.amazonaws.com`,
        apiVersion: "2011-06-15",
        region: config.region,
        accessKeyId: config.accessKeyId,
        secretAccessKey: config.secretAccessKey
    });

    const stsAssecume = {
        DurationSeconds: sessionDuration,//defalut 1H min 15min
        RoleSessionName: "loop", //can be any value
        RoleArn: roleArn
    };
    console.log(`[ STS ] Utils sts.assumeRole`);
    return sts.assumeRole(stsAssecume).promise();
}

async function prepareESPCredentails(resource, sessionDuration) {
    let params = {
        region: config.region,
        durationInSeconds: sessionDuration
    };
    if (resource === AWSRESOURCE.S3) {
        params.iamRoleArn = config.s3RoleArn;
        params.externalId = config.s3ExternalId
    } else if (resource === AWSRESOURCE.SQS) {
        params.iamRoleArn = config.sqsRoleArn;
        params.externalId = config.sqsExternalId
    }

    let url = config.espTokenAuthorizeUrl;
    const data = qs.stringify({ token: config.espToken });
    let configs = { headers: { "Content-Type": "application/x-www-form-urlencoded" } }
    const espToken = await axios.post(url, data, configs);

    url = config.espAwsAuthorizeUrl
    configs = { params: params, headers: { 'Authorization': espToken.data.access_token } }
    const res = await axios.get(url, configs);

    const Credentials = {
        AccessKeyId: res.data.accessKeyId,
        SecretAccessKey: res.data.secretAccessKey,
        SessionToken: res.data.sessionToken
    }
    return { Credentials }

}

module.exports = {
    prepareCredentials,
    prepareOSCredentails,
    prepareESPCredentails,

}

File: active-learning-service/src/aws/sts.py
# Copyright 2019-2021 VMware, Inc.
# SPDX-License-Identifier: Apache-2.0


import boto3
import requests
import json

from config.config import config


def sts_role(arn, session_name):

    sts = boto3.client(
        'sts',
        region_name=config["REGION"],
        aws_access_key_id=config["AWS_ACCESS_KEY_ID"],
        aws_secret_access_key=config["AWS_SECRET_ACCESS_KEY"]
    )

    data = sts.assume_role(
        DurationSeconds=1800,
        RoleArn=arn,
        RoleSessionName=session_name
    )
    data = data['Credentials']

    return {
        'accessKeyId': data['AccessKeyId'],
        'secretAccessKey': data['SecretAccessKey'],
        'sessionToken': data['SessionToken']
    }


def esp_sts_credentials(token):
    data = {
        "iamRoleArn": config["S3_ROLE_ARN"],
        "externalId": config["S3_EXTERNAL"],
        "region": config["REGION"],
        "durationInSeconds": 1800
    }
    headers = {"Authorization": token}

    credentials = requests.get(config["ESP_AUTHORIZE_URL"], params=data, headers=headers)
    credentials = json.loads(credentials.content)

    return credentials


def aws_credentials(token):
    if "ESP" in config and config["ESP"]:
        credentials = esp_sts_credentials(token)
    else:
        credentials = sts_role(config["S3_ROLE_ARN"], 'loop_al')

    return credentials
